make throttling optional and configurable. fix #2

andreineculau opened the pull request:
NOTE: I ran the performance checks again, and the numbers seem much better than before. We could try using no throttling and see the performance 💣

IanSavchenko commented on src/util.js:
```diff
@@ -23,8 +22,25 @@ export let roundRect = function(obj, precision) {
+  };
+};
+
+export let throttleDeeply = function(obj, wait) {
```
Would be great if we can avoid throttling!

But I don't understand why it can't be done simpler? Check in `throttle` if `cfg.throttle` is 0 or `undefined` and return function as it is. Otherwise, wrap it into `throttle`. 3 extra lines and that's all or am I missing something?

It took me half an hour to understand what is going on here and another half an hour to test it and find couple bugs. Don't know how much time it took you to rename all the usages...

andreineculau replied:
The answer to your question is a question: when? when should `throttle` check? because as it is now (pre-PR) `throttle` is called whenever the module is parsed (i.e. today the `cfg.throttle` is "broken", or to be politically correct can only be changed in code, not at runtime).

IanSavchenko commented on src/util.js:
```diff
@@ -23,8 +22,25 @@ export let roundRect = function(obj, precision) {
+  return fn;
+};
+
+export let deeply = function(fn) {
```
Couple things:
1) There is no function `_.deeply`
2) `fn` is invoked for array of values, not a single plain object value. 

Something like this works for me:
```
let deeply = function(fn) {
  return function(obj, iteratee) {
    return _.mapValues(obj, function(v, _name) {
      return _.isPlainObject(v) ? deeply(fn)(v, iteratee) : fn(v, iteratee);
    });
  };
};
```

IanSavchenko reviewed (changes requested):
Generally, I support this initiative, but this solution looks to me as too complicated and error-prone.

Commit:
make throttling optional and configurable. fix #2

andreineculau commented:
Fixed `deeply` and updated the PR. Thanks for the input! FWIW I didn't test it when I opened the PR :) I wanted to discuss it on the next working day, which never happened...

andreineculau commented:
TBD we test without any throttling, and if all is still fine, we just remove the throttling altogether in an updated version of this PR

Commit:
sort

Commit:
add top level toJSON function

andreineculau commented:
🤷‍♂️
